content/blog/2023-11-23-announcing-axum-0-7-0.md

```diff
@@ -0,0 +1,187 @@
+
+There isn't yet a full replacement in `hyper-util` so `axum` now provides its
+own `serve` function:
+
+```rust
+use axum::{
+    routing::get,
+    Router,
+};
+use tokio::net::TcpListener;
+
+let app = Router::new().route("/", get(|| async { "Hello, World!" }));
+
+let listener = TcpListener::bind("0.0.0.0:3000").await.unwrap();
+axum::serve(listener, app).await.unwrap();
+```
+
+The purpose of `axum::serve` is to provide a way to get started with `axum`
+quickly and as such it does not support any configuration options whatsoever.
+If you need configuration you have to use `hyper` and `hyper-util` directly. We
+provide an example showing how to do that [here][hyper-serve-example].
+
+## Our own `Body` type
+
+The [`http-body`] crate is now also at 1.0 and that comes with a similar API split
+that `hyper` and `hyper-util` have. `http-body` now just provides the core APIs,
+and high level utilities has been moved to [`http-body-util`]. That includes
+things like `Full`, `Empty`, and `UnsyncBoxBody`, which used to be re-exported by
+`axum`.
+
+For the same reason that `hyper-util` shouldn't be part of `axum`'s public API,
+`http-body-util` shouldn't either.
+
+As a replacement, `axum` now provides its own body type found at `axum::body::Body`.
+
+Its API is similar to what `hyper::Body` had:
+
+```rust
+use axum::body::Body;
+use futures::TryStreamExt;
+use http_body_util::BodyExt;
+
+// Create an empty body (Body::default() does the same)
+Body::empty();
+
+// Create bodies from strings or buffers
+Body::from("foobar");
+Body::from(Vec::<u8>::from([1, 3, 3, 7]));
+
+// Wrap another type that implements `http_body::Body`
+Body::new(another_http_body);
+
+// Convert a `Body` into a `Stream` of data frames
+let mut stream = body.into_data_stream();
+while let Some(chunk) = stream.try_next().await? {
+    // ...
+}
+
+// Collect the body into a `Bytes`. Uses `BodyExt::collect`
+// This replaces the previous `hyper::body::to_bytes` function
+let bytes = body.collect().await?.to_bytes();
+```
+
+## Fewer generics
+
+`axum::Router` used to be generic over the request body type. That meant
+applying middleware that changed the request body type would have knock-on
+effects throughout your routes:
+
+```rust
+// This would work just fine
+Router::new()
+    .route(
+        "/",
+        get(|body: Request<Body>| async { ... })
+    );
+
+// But adding `tower_http::limit::RequestBodyLimitLayer` would make
+// things no longer compile
+Router::new()
+    .route(
+        "/",
+        get(|body: Request<Body>| async { ... })
+    )
+    .layer(tower_http::limit::RequestBodyLimitLayer::new(1024));
+```
+
+The reason it doesn't work is that `RequestBodyLimitLayer` changes the request
+body type so you have to extract `Request<http_body::Limited<Body>>` instead of
+`Request<Body>`. This was very subtle and the source of some confusion.
+
+In axum 0.7, everything continues working as before, regardless of which
+middleware you add:
```

Comment:
I assume this is because we now use an internally-boxed dynamic dispatch type for `Body`? Is that worth calling out?

Reply:
Exactly. `axum::Router` implements `Service<Request<B>>` for any `B` and then immediately boxes the body in `Service::call` https://docs.rs/axum/latest/src/axum/routing/mod.rs.html#450-469

I'll add a comment about that.